memchunk_tpl.h

```diff
@@ -0,0 +1,166 @@
+/* memchunk_tpl.h -- inline functions to copy small data chunks.
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+/* Returns the chunk size */
+uint32_t CHUNKSIZE(void) {
+    return sizeof(memchunk_t);
+}
+
+/* Behave like memcpy, but assume that it's OK to overwrite at least
+   memchunk_t bytes of output even if the length is shorter than this,
+   that the length is non-zero, and that `from` lags `out` by at least
+   sizeof memchunk_t bytes (or that they don't overlap at all or simply that
+   the distance is less than the length of the copy).
+
+   Aside from better memory bus utilisation, this means that short copies
+   (memchunk_t bytes or fewer) will fall straight through the loop
+   without iteration, which will hopefully make the branch prediction more
+   reliable. */
+uint8_t* CHUNKCOPY(uint8_t *out, uint8_t const *from, unsigned len) {
+    memchunk_t chunk;
+    --len;
+    loadchunk(from, &chunk);
+    storechunk(out, &chunk);
+    out += (len % sizeof(memchunk_t)) + 1;
+    from += (len % sizeof(memchunk_t)) + 1;
+    len /= sizeof(memchunk_t);
+    while (len > 0) {
+        loadchunk(from, &chunk);
+        storechunk(out, &chunk);
+        out += sizeof(memchunk_t);
+        from += sizeof(memchunk_t);
+        --len;
+    }
+    return out;
+}
+
+/* Behave like chunkcopy, but avoid writing beyond of legal output. */
+uint8_t* CHUNKCOPY_SAFE(uint8_t *out, uint8_t const *from, unsigned len, uint8_t *safe) {
+    if ((safe - out) < (ptrdiff_t)sizeof(memchunk_t)) {
+        if (len & 8) {
```

Comment:
Here, len modulo memchunk_t can be more than 7, when using 256-bit vector.

Reply:
Ah you are probably right.

Reply:
You can still make another pull request with all the suggestions not currently on this one... Makes easier to review ;)